src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php

```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
             $aliasId = (string) $alias;
-
+            if (isset($replacements[$aliasId])) {
```

Comment:
What if you have chained aliased, and the target alias is processed only later ? This would not replace it properly anymore (which makes me think that this case is not properly covered by tests btw)